man/man1/mkvh60.1
.TH MKVH60 1NEMO "19 April 2021"

.SH "NAME"
mkvh60 \- create a uniform sphere of equal mass stars using von Hoerner's 1960 recipe

.SH "SYNOPSIS"
\fBmkvh60\fP [parameter=value]

.SH "DESCRIPTION"
Sebastian von Hoerner (1960ZA.....50..184V) describes - in beautiful German -
four steps how his equal mass virialized homogenous
sphere was constructed. It also uses a random number generator he himself designed. Here's my interpretation
of these steps:
.RS 5
.IP 1.
Get random numbers between 0 and 1 (see ran_svh57(3NEMO)).
.IP 2.
Ensure positions and velocities are uniformly distributed
in a sphere of radius 1 (in both x,y,z and vx,vy,vz)
.IP 3.
Shift center of mass to exactly 0.0
.IP 4.
Rescale velocities such that the virial (2Ek/Ep) is exactly 1.0
.RE

Note that the velocities would not be gaussian if my interpretation is correct (cf. his eq.(22) and (71)), thus
the initial conditions are not in the final phase space configuration, yet the models are fun to play with.
.PP
Since the simulations were done on the Siemens 2002 computer, with about 2000 FLOPS,
the number of particles was limited to 16 in his 1960 paper, though his followup paper in 1963
discussed the case nbody=25!
.PP
The seed is special in von Hoerner. It is a random number between 0.57 and 0.91, and uses is own 1957 method,
implemented in \fIran_svh57(3NEMO)\fP. By default this code will use NEMO's default random numbers,
set by \fBseed=\fP, however if \fBseed57\fP is given, his code is used. See also \fIxrandom(1NEMO)\fP.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBout=\fP
Output file name, in standard \fIsnapshot(5NEMO)\fP format. No default.
.TP
\fBnbody=\fP
Number of particles [16]    
.TP
\fBseed57=\fP
Von Hoerners 1957 seed, must be (0.57 - 0.91). Default: not used.
.TP
\fBseed=\fP
NEMO's Random number seed, if used. [0] 
.TP
\fBzerocm=\fP
Center c.o.m. ? [t]    
.TP
\fBheadline=\fP
Text headline for output. By default not used.

.SH "EXAMPLES"
Here is a default n=16 sphere, first using the NEMO random number generator, and then the von Hoerner ran57 method:

.nf
\fBmkvh60 . 16 seed=123\fP
### nemo Debug Info: virial radius: 0.740663    v2=0.821627   ek=0.218607 ep=0.632879   K=1.44753
### nemo Debug Info: snapvscale: 1.20313

\fBmkvh60 . 16 seed57=0.580128\fP
### nemo Debug Info: SvH seed=0.580128
### nemo Debug Info: von Hoerner seed57=0.580128 first ran=0.548496
### nemo Debug Info: virial radius: 0.759762    v2=0.811234   ek=0.261934 ep=0.616969   K=1.17772
### nemo Debug Info: snapvscale: 1.08523

.fi

.SH "SEED57"
The \fBseed57=\fP keyword is the seed for the revived ran_svh57() \fIZufallzahl\fP random number generator.
My interpretation of eq.(44) in 1960ZA.....50..184V is : (code in python)

.nf
  x = seed57      # between 0.57 and 0.91
  while True:
    x2 = x*x
    if x2 < xlo:
      xnew = x2 + 0.32
    else:
      xnew = x2
    ran = int(('%.12f' % xnew)[5:])/1e9     # print out with 9 digits
    print(ran)
    x = xnew
.fi    

.SH "SEE ALSO"
mkhomsph(1NEMO), firstn(1NEMO), snapshot(1NEMO)
.PP
.nf
Z.f.Astrophys. 50, 184, 1960  (1960ZA.....50..184V)
2001ASPC..228...11V  \fIHow it all started\fP
Siemens 2002 computer: https://dl.acm.org/doi/10.1145/1458043.1458076
.fi

.SH "FILES"
src/nbody/init

.SH "AUTHOR"
Sebastian von Hoerner (1960), Peter Teuben (NEMO re-implementation)

.SH "HISTORY"
.nf
.ta +1.5i +5.5i
19-Apr-2021	V0.1 Created	PJT
.fi


src/nbody/init/mkvh60.c
/*
 * MKVH60: set up a uniform sphere with equal mass particles
 *         according to von Hoerner's (1960) recipe - 1960ZA.....50..184V
 *	
 *         18-apr-2021  drafted                  PJT
 */


#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/put_snap.c>

string defv[] = {	/* DEFAULT INPUT PARAMETERS */
    "out=???\n      Output file name",
    "nbody=10\n     Number of particles",
    "seed57=\n      Von Hoerners 1957 seed, must be (0.57 - 0.91)",
    "seed=0\n       NEMO's Random number seed (if used)",
    "zerocm=t\n     Center c.o.m. ?",
    "headline=\n    Text headline for output",
    "VERSION=0.2\n  19-apr-2021 PJT",
    NULL,
};

string usage = "create a uniform sphere of equal mass stars using von Hoerner's 1960 recipe";

string cvsid="$Id$";


local real virial=1, power=0;
local bool zerocm;
local bool exact = TRUE;
local bool Q57;

local Body *btab;
local int nbody;

extern double xrandom(double,double), grandom(double,double);
extern double ran_svh57(double);

void writegalaxy(string name, string headline);
void mksphere(void);
void centersnap(body *btab, int nb);
void snapvscale(body *btab, int nb, double vscale);
double get_vr(body *btab, int nb);
double get_ek(Body *btab, int nb);

void nemo_main(void)
{
    int seed;
    real seed57;

    nbody = getiparam("nbody");
    seed = init_xrandom(getparam("seed"));
    zerocm = getbparam("zerocm");
    Q57 = hasvalue("seed57");
    if (Q57) {
      seed57 = getrparam("seed57");
      dprintf(0,"von Hoerner seed57=%g first ran=%g\n",seed57,ran_svh57(seed57));
    }
    mksphere();
    writegalaxy(getparam("out"), getparam("headline"));
}

/*
 * WRITEGALAXY: write galaxy model to output.
 */

void writegalaxy(string name, string headline)
{
    stream outstr;
    real tsnap = 0.0;
    int bits = MassBit | PhaseSpaceBit | TimeBit;

    if (! streq(headline, ""))
	set_headline(headline);
    outstr = stropen(name, "w");
    put_history(outstr);
    put_snap(outstr, &btab, &nbody, &tsnap, &bits);
    strclose(outstr);
}

/*
 * MKSPHERE: homogeneous sphere
 */

void mksphere(void)
{
    Body *bp;
    real r_i, v_i, theta_i, phi_i, mass_i, sigma = 0.0;
    real sinp, cosp, sint, cost;
    real rmax = 1.0;
    real ot = 1.0/3.0;
    int i;

    btab = (Body *) allocate(nbody * sizeof(Body));
    mass_i = 1.0/nbody;
    if (virial > 0.0) {
        /* the sphere has potential energy -3GM^2/(5R)  */
        /* sigma is the 1D velocity dispersion of an isotropic one */
        sigma = sqrt(virial/(5*rmax));
        dprintf(1,"Gaussian isotropic velocities: sigma=%g\n",sigma);
    } 
    for (bp=btab, i = 0; i < nbody; bp++, i++) {
	Mass(bp) = mass_i;

	if (Q57) {
	  // positions
	  r_i = pow(ran_svh57(0.0),ot);
	  theta_i = acos(2.0*ran_svh57(0.0) - 1.0);
	  phi_i = ran_svh57(0.0) * TWO_PI;
	  Phase(bp)[0][0] = r_i * sin(theta_i) * cos(phi_i);
	  Phase(bp)[0][1] = r_i * sin(theta_i) * sin(phi_i);
	  Phase(bp)[0][2] = r_i * cos(theta_i);

	  // velocities, same procedure as positions
	  r_i = pow(ran_svh57(0.0),ot);
	  theta_i = acos(2.0*ran_svh57(0.0) - 1.0);
	  phi_i = ran_svh57(0.0) * TWO_PI;
	  Phase(bp)[1][0] = r_i * sin(theta_i) * cos(phi_i);
	  Phase(bp)[1][1] = r_i * sin(theta_i) * sin(phi_i);
	  Phase(bp)[1][2] = r_i * cos(theta_i);

	} else {
	  // positions
	  r_i = pow(xrandom(0.0,1.0),ot);
	  theta_i = acos(xrandom(-1.0,1.0));
	  phi_i = xrandom(0.0,TWO_PI);
	  Phase(bp)[0][0] = r_i * sin(theta_i) * cos(phi_i);
	  Phase(bp)[0][1] = r_i * sin(theta_i) * sin(phi_i);
	  Phase(bp)[0][2] = r_i * cos(theta_i);

	  // velocities, same procedure as positions
	  r_i = pow(xrandom(0.0,1.0),ot);
	  theta_i = acos(xrandom(-1.0,1.0));
	  phi_i = xrandom(0.0,TWO_PI);
	  Phase(bp)[1][0] = r_i * sin(theta_i) * cos(phi_i);
	  Phase(bp)[1][1] = r_i * sin(theta_i) * sin(phi_i);
	  Phase(bp)[1][2] = r_i * cos(theta_i);
	}
	
    }
    // center the snapshot (you almost never want to skip this)
    if (zerocm)
        centersnap(btab,nbody);

    // rescale so it's in global virial equilibrium
    double rv = get_vr(btab,nbody);
    double v2 = 1/sqrt(2*rv);
    double ek = get_ek(btab,nbody);
    double ep = 0.5*(nbody-1)/(nbody*rv);
    double K = ep/(2*ek);
    dprintf(0,"virial radius: %g    v2=%g   ek=%g ep=%g   K=%g\n",rv,v2,ek,ep,K);
    snapvscale(btab, nbody, sqrt(K));
    
}


void centersnap(Body *btab, int nb)
{
    real mtot;
    vector cmphase[2], tmp;
    Body *bp;

    mtot = 0.0;
    CLRV(cmphase[0]);
    CLRV(cmphase[1]);
    for (bp = btab; bp < btab + nb; bp++) {
	mtot = mtot + Mass(bp);
	MULVS(tmp, Phase(bp)[0], Mass(bp));
	ADDV(cmphase[0], cmphase[0], tmp);
	MULVS(tmp, Phase(bp)[1], Mass(bp));
	ADDV(cmphase[1], cmphase[1], tmp);
    }
    MULVS(cmphase[0], cmphase[0], 1.0/mtot);
    MULVS(cmphase[1], cmphase[1], 1.0/mtot);
    for (bp = btab; bp < btab + nb; bp++) {
	SUBV(Phase(bp)[0], Phase(bp)[0], cmphase[0]);
	SUBV(Phase(bp)[1], Phase(bp)[1], cmphase[1]);
    }
}

void snapvscale(Body *btab, int nb, real vscale)
{
    real mtot;
    vector cmphase[2], tmp;
    Body *bp;

    dprintf(0,"snapvscale: %g\n",vscale);

    for (bp = btab; bp < btab + nb; bp++)
      SMULVS(Phase(bp)[1], vscale);
}

double get_vr(Body *btab, int nb)
{
  real sum;
  Body *bp, *qp;

  sum = 0.0;
  for (bp = btab; bp < btab + nb; bp++) {
    for (qp = bp+1; qp < btab + nb; qp++) {
      sum += 1.0/distv(Phase(bp)[0], Phase(qp)[0]);
    }
  }
  sum *= (2.0) / (nb*(nb-1));
  return 1.0/sum;
}

double get_ek(Body *btab, int nb)
{
  real sum;
  Body *bp;

  sum = 0.0;
  for (bp = btab; bp < btab + nb; bp++)
    sum += dotvp(Phase(bp)[1], Phase(bp)[1]);
  return sum/2.0/nb;
}

